### PR TITLE
*: reduce sys_getdents syscall

### PR DESCRIPTION
**Pull request by breeswish**
### What problem does this PR solve?

This PR reduce the number of syscalls to `sys_getdents`. `sys_getdents` comes from `read_dir`, causes many slab allocations if the directory to scan contains too many files. In our case, TiKV does two kind of heavy directory scans:

- `/proc/PID/fd` to know how many open fds. This will scan > 50K files in common production environment.
- `/proc/PID/task` to know active threads. This will scan ~500 files.

 If TiKV calls `sys_getdents` frequently, it will cause slab fragmentations when TiKV server is running for long time. When slab fragmentations are notable, the performance of whole system will be slowed down, thus affect the service.

A linux kernel RFC demostates this fragmentation issue:

image

### What is changed and how it works?

1. Removing the `process` feature from rust-prometheus, which includes an active fd collector that calls `sys_getdents` when reading dir `/proc/PID/fd`.

2. ~~Enlarge the pool eagerly at creation time by appending dummy tasks. In this way, there will be no more new dynamic worker threads.~~ We use yatp now, so there is no new dynamic worker threads as well.

3. Thanks to (2), we can safely avoid frequently updating thread list in the PD collector, since all threads are static now. The thread list is updated by interval 15s, 30s, 1min, ..., 10min if there is no thread changes.

4. #5598 is partially reverted. There is a bug in #5598 that actually it calculates the thread load according to all threads at initial time (instead of normal read pool threads). Since now all threads are static in (2), the request batcher now calculates based on correct threads, at creation time.

### Related changes

- Need to cherry-pick to the release branch: 4.0, 3.1, 3.0, 2.1, 2.0.

### Check List 

Tests 

- Manual test using `strace -f -s 1024 -tt -T -yy -p $PID -e getdents`

current master (~124 calls in 5 minutes):

<details>

```
[pid 120734] 15:13:50.919711 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000109>
[pid 120734] 15:13:50.920100 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000019>
[pid 120734] 15:13:55.967730 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000118>
[pid 120734] 15:13:55.968033 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000018>
[pid 120734] 15:14:01.025810 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000099>
[pid 120734] 15:14:01.026068 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000019>
[pid 120734] 15:14:06.075101 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000188>
[pid 120734] 15:14:06.075454 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:14:11.129743 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000197>
[pid 120734] 15:14:11.130136 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:14:16.178394 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000159>
[pid 120734] 15:14:16.178699 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000022>
[pid 120734] 15:14:21.233495 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000148>
[pid 120734] 15:14:21.233865 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:14:26.284165 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000104>
[pid 120734] 15:14:26.284443 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000013>
[pid 120734] 15:14:31.337966 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000110>
[pid 120734] 15:14:31.338231 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000019>
[pid 120725] 15:14:33.857266 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 2 entries */, 32768) = 48 <0.000047>
[pid 120725] 15:14:33.857434 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:14:36.388340 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000155>
[pid 120734] 15:14:36.388699 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000047>
[pid 120734] 15:14:41.445010 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000105>
[pid 120734] 15:14:41.445327 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:14:46.497561 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000109>
[pid 120734] 15:14:46.497837 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000018>
[pid 120734] 15:14:51.550864 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000123>
[pid 120734] 15:14:51.551150 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000037>
[pid 120734] 15:14:56.622597 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000277>
[pid 120734] 15:14:56.623029 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:15:01.671998 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000185>
[pid 120734] 15:15:01.672360 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:15:06.737462 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000110>
[pid 120734] 15:15:06.737723 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000019>
[pid 120734] 15:15:11.782398 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000283>
[pid 120734] 15:15:11.782917 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000028>
[pid 120734] 15:15:16.843735 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000276>
[pid 120734] 15:15:16.844208 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000024>
[pid 120734] 15:15:21.893342 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000225>
[pid 120734] 15:15:21.893713 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000015>
[pid 120734] 15:15:26.950145 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000211>
[pid 120734] 15:15:26.950524 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:15:32.019030 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000239>
[pid 120734] 15:15:32.019423 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120725] 15:15:33.859471 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 2 entries */, 32768) = 48 <0.000033>
[pid 120725] 15:15:33.859608 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:15:37.072722 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000201>
[pid 120734] 15:15:37.073077 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:15:42.127232 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000158>
[pid 120734] 15:15:42.127567 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000019>
[pid 120734] 15:15:47.199325 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000141>
[pid 120734] 15:15:47.199605 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000018>
[pid 120734] 15:15:52.257264 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000188>
[pid 120734] 15:15:52.257701 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000038>
[pid 120734] 15:15:57.307516 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000174>
[pid 120734] 15:15:57.307912 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:16:02.378290 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000168>
[pid 120734] 15:16:02.378622 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:16:07.445609 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000128>
[pid 120734] 15:16:07.445884 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000014>
[pid 120734] 15:16:12.500156 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000184>
[pid 120734] 15:16:12.500489 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:16:17.569908 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000177>
[pid 120734] 15:16:17.570296 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:16:22.638112 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000181>
[pid 120734] 15:16:22.638444 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000014>
[pid 120734] 15:16:27.690347 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000234>
[pid 120734] 15:16:27.690755 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000019>
[pid 120734] 15:16:32.749602 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000123>
[pid 120734] 15:16:32.749859 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000017>
[pid 120725] 15:16:33.861692 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 2 entries */, 32768) = 48 <0.000054>
[pid 120725] 15:16:33.861837 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:16:37.797510 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000163>
[pid 120734] 15:16:37.797839 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000022>
[pid 120734] 15:16:42.860393 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000205>
[pid 120734] 15:16:42.860819 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000027>
[pid 120734] 15:16:47.921781 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000171>
[pid 120734] 15:16:47.922142 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000018>
[pid 120734] 15:16:52.977924 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000212>
[pid 120734] 15:16:52.978303 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:16:58.046162 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000168>
[pid 120734] 15:16:58.046529 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000049>
[pid 120734] 15:17:03.108488 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000198>
[pid 120734] 15:17:03.108874 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:17:08.164334 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000180>
[pid 120734] 15:17:08.164728 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000027>
[pid 120734] 15:17:13.226356 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000183>
[pid 120734] 15:17:13.226686 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:17:18.295925 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000214>
[pid 120734] 15:17:18.296320 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:17:23.370720 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000194>
[pid 120734] 15:17:23.371099 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:17:28.437188 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000174>
[pid 120734] 15:17:28.437516 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000030>
[pid 120734] 15:17:33.487587 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000197>
[pid 120734] 15:17:33.487952 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000015>
[pid 120724] 15:17:33.863881 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 2 entries */, 32768) = 48 <0.000050>
[pid 120724] 15:17:33.864039 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 0 entries */, 32768) = 0 <0.000022>
[pid 120734] 15:17:38.545227 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000152>
[pid 120734] 15:17:38.545569 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:17:43.597446 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000163>
[pid 120734] 15:17:43.597764 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:17:48.664991 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000215>
[pid 120734] 15:17:48.665375 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000022>
[pid 120734] 15:17:53.722447 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000195>
[pid 120734] 15:17:53.722833 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000026>
[pid 120734] 15:17:58.796483 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000184>
[pid 120734] 15:17:58.796826 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000020>
[pid 120734] 15:18:03.853628 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000187>
[pid 120734] 15:18:03.854021 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000013>
[pid 120734] 15:18:08.908492 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000116>
[pid 120734] 15:18:08.908739 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000019>
[pid 120734] 15:18:13.963686 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000125>
[pid 120734] 15:18:13.963944 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000014>
[pid 120734] 15:18:19.013452 getdents(533</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000250>
[pid 120734] 15:18:19.013883 getdents(533</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000021>
[pid 120734] 15:18:24.082390 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000190>
[pid 120734] 15:18:24.082770 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000014>
[pid 120734] 15:18:29.138893 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000173>
[pid 120734] 15:18:29.139233 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000022>
[pid 120725] 15:18:33.866065 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 2 entries */, 32768) = 48 <0.000030>
[pid 120725] 15:18:33.866202 getdents(10</data3/wenxuan_kv1_master/data/snap>, /* 0 entries */, 32768) = 0 <0.000030>
[pid 120734] 15:18:34.189182 getdents(10</proc/120082/task>, /* 170 entries */, 32768) = 5424 <0.000150>
[pid 120734] 15:18:34.189486 getdents(10</proc/120082/task>, /* 0 entries */, 32768) = 0 <0.000015>
```

</details>

this PR (20 calls in 5 minutes, there will be fewer then):

<details>

```
[pid 118990] 15:13:35.627680 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 2 entries */, 32768) = 48 <0.000042>
[pid 118990] 15:13:35.627949 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 0 entries */, 32768) = 0 <0.000022>
[pid 118999] 15:13:52.946732 getdents(558</proc/118262/task>, /* 188 entries */, 32768) = 6000 <0.000172>
[pid 118999] 15:13:52.947348 getdents(558</proc/118262/task>, /* 0 entries */, 32768) = 0 <0.000022>
[pid 119044] 15:14:01.629937 getdents(558</proc/118262/task>, /* 188 entries */, 32768) = 6000 <0.000149>
[pid 119044] 15:14:01.630460 getdents(558</proc/118262/task>, /* 0 entries */, 32768) = 0 <0.000027>
[pid 118989] 15:14:35.629657 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 2 entries */, 32768) = 48 <0.000026>
[pid 118989] 15:14:35.629807 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 0 entries */, 32768) = 0 <0.000024>
[pid 118990] 15:15:35.631421 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 2 entries */, 32768) = 48 <0.000056>
[pid 118990] 15:15:35.631638 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 0 entries */, 32768) = 0 <0.000041>
[pid 118990] 15:16:35.633434 getdents(553</data3/wenxuan_kv1_exp/data/snap>, /* 2 entries */, 32768) = 48 <0.000033>
[pid 118990] 15:16:35.633621 getdents(553</data3/wenxuan_kv1_exp/data/snap>, /* 0 entries */, 32768) = 0 <0.000041>
[pid 118989] 15:17:35.635635 getdents(557</data3/wenxuan_kv1_exp/data/snap>, /* 2 entries */, 32768) = 48 <0.000041>
[pid 118989] 15:17:35.635819 getdents(557</data3/wenxuan_kv1_exp/data/snap>, /* 0 entries */, 32768) = 0 <0.000058>
[pid 118999] 15:17:52.596959 getdents(521</proc/118262/task>, /* 188 entries */, 32768) = 6000 <0.000162>
[pid 118999] 15:17:52.597600 getdents(521</proc/118262/task>, /* 0 entries */, 32768) = 0 <0.000036>
[pid 119044] 15:18:01.626498 getdents(521</proc/118262/task>, /* 188 entries */, 32768) = 6000 <0.000147>
[pid 119044] 15:18:01.627047 getdents(521</proc/118262/task>, /* 0 entries */, 32768) = 0 <0.000062>
[pid 118989] 15:18:35.637626 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 2 entries */, 32768) = 48 <0.000035>
[pid 118989] 15:18:35.637771 getdents(558</data3/wenxuan_kv1_exp/data/snap>, /* 0 entries */, 32768) = 0 <0.000024>
```

</details>

### Release note 

- Fixed a bug that causes notable memory fragmentation when TiKV node is running for long time

### CODE REVIEW

**Comment by BusyJay** [comment]
How about using yatp future pool to replace tokio thread pool?

**Comment by siddontang** [comment]
what real benefits can we get from reducing `sys_getdents`? can you show us here?

**Comment by breeswish** [comment]
@siddontang Thanks for the suggestion! Updated the issue body as below:

This PR reduce the number of syscalls to sys_getdents. sys_getdents comes from read_dir, causes many slab allocations if the directory to scan contains too many files. In our case, TiKV does two kind of heavy directory scans:

- /proc/PID/fd to know how many open fds. This will scan > 50K files in common production environment.
- /proc/PID/task to know active threads. This will scan ~500 files.

If TiKV calls sys_getdents frequently, it will cause slab fragmentations when TiKV server is running for long time. When slab fragmentations are notable, the performance of whole system will be slowed down, thus affect the service.

@BusyJay yatp is gated behind a flag considering its stability. Its fallback solution is the current tokio-threadpool. Thus I think yatp may not be appropriate to be used to replace the fallback solution itself.

**Comment by BusyJay** [comment]
The flag is used to opt out multi-level feedback queue. yatp future pool only use single level, doesn't have such feature.

**Comment by BusyJay** [comment]
> Fixed a bug that causes notable memory fragmentation when TiKV node is running for long time

Any issue or memory graph that demonstrate the problem and the  cause?

The example in the issue seems different from our cases. Because the directory is listed frequently, so memory should be often reclaimed instead of freed randomly. /cc @ethercflow

**Comment by breeswish** [comment]
> The flag is used to opt out multi-level feedback queue. yatp future pool only use single level, doesn't have such feature.

Nice to hear. @sticnarf How do you think?

**Comment by sticnarf** [comment]
> > The flag is used to opt out multi-level feedback queue. yatp future pool only use single level, doesn't have such feature.
> 
> Nice to hear. @sticnarf How do you think?

Yes, I think we can replace tikv_util/future_pool with yatp's future pool with single level.

File: components/tikv_util/src/metrics/threads_linux.rs [hunk 1]
```diff
@@ -109,8 +111,9 @@ impl Collector for ThreadsCollector {
         // Clean previous threads state.
         metrics.threads_state.reset();
 
-        let tids = get_thread_ids(self.pid).unwrap();
-        for tid in tids {
+        let mut tid_retriever = self.tid_retriever.lock().unwrap();
+        let tids = tid_retriever.get_tids();
+        for &tid in tids {
```
Comment (BusyJay) [inline components/tikv_util/src/metrics/threads_linux.rs:116]: `*tid` is a preferred style.

File: src/server/load_statistics/linux.rs [hunk 2]
```diff
@@ -31,25 +29,28 @@ impl ThreadLoadStatistics {
     /// Note: call this after the target threads are initialized, otherwise it can't catch them.
     pub fn new(slots: usize, prefix: &str, thread_load: Arc<ThreadLoad>) -> Self {
         let pid: pid_t = unsafe { getpid() };
-        let (tids, cpu_total_count) = poll_task(pid, prefix);
+        let mut tids = vec![];
```
Comment (BusyJay) [inline src/server/load_statistics/linux.rs:32]: @hicqu PTAL

File: components/tikv_util/src/metrics/memory_linux.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    descs: Vec<Desc>,
+    vsize: Gauge,
+    rss: Gauge,
+}
+
+impl MemoryCollector {
+    pub fn new(pid: libc::pid_t) -> Self {
+        let mut descs = Vec::new();
+
+        let vsize = Gauge::with_opts(Opts::new(
```
Comment (BusyJay) [inline components/tikv_util/src/metrics/memory_linux.rs:33]: Can it be solved by node exporter?

Comment (breeswish) [reply components/tikv_util/src/metrics/memory_linux.rs:33]: Node exporter exports the memory usage of node exporter itself and looks like there is [no config](https://prometheus.io/docs/guides/node-exporter/) for the node exporter.

File: components/tikv_util/src/metrics/memory_linux.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+use std::io::{Error, ErrorKind, Result};
+
+use libc;
+use procinfo::pid as pid_info;
+
+use prometheus::core::{Collector, Desc};
+use prometheus::{proto, Gauge, Opts};
+
+/// Monitors threads of the current process.
```
Comment (sticnarf) [inline components/tikv_util/src/metrics/memory_linux.rs:14]: The comment does not seem right.

File: components/tikv_util/src/metrics/memory_linux.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+}
+
+impl Collector for MemoryCollector {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        if let Ok(statm) = pid_info::statm(self.pid) {
```
Comment (sticnarf) [inline components/tikv_util/src/metrics/memory_linux.rs:62]: Is there any special consideration of calling `statm` with pid instead of `statm_self`?

Comment (breeswish) [reply components/tikv_util/src/metrics/memory_linux.rs:62]: nop, it's simply copied from prometheus.

File: components/tikv_util/src/metrics/threads_linux.rs (hunk 1 above)
Comment (breeswish) [reply components/tikv_util/src/metrics/threads_linux.rs:116]: Do you mean:

```rust
for tid in tids {
    let tid = *tid;
    ...
}
```

Comment (BusyJay) [reply components/tikv_util/src/metrics/threads_linux.rs:116]: Yes.

**Pull request merged**